## __Data Cleaning the Restaurant Data__

In [ ]:
# import key libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### __Loading data from zomato.csv__

In [120]:
df = pd.read_csv('zomato.csv')
df.head()

url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone  \
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

### Get the shape of the DataFrame: Rows and columns

In [121]:
df.shape

(51717, 17)

### Check columns

In [122]:
df.columns

Index(['url', 'address', 'name', 'online_order', 'book_table', 'rate', 'votes',
       'phone', 'location', 'rest_type', 'dish_liked', 'cuisines',
       'approx_cost(for two people)', 'reviews_list', 'menu_item',
       'listed_in(type)', 'listed_in(city)'],
      dtype='object')

### Drop redundant/unnecessary columns

In [123]:
# Remove url, address, phone, dish_liked, reviews_list, menu_item
df = df.drop(['url', 'address', 'phone', 'dish_liked', 'reviews_list', 'menu_item'], axis=1)

### Drop any duplicate rows

In [124]:
df.shape

(51717, 11)

In [125]:
df.drop_duplicates(inplace=True)

In [126]:
df.shape

(51609, 11)

### Renaming columns
- __First inspect columns to rename with df.columns command__
- __Then delete similar columns__

In [127]:
df.columns.tolist()

['name',
 'online_order',
 'book_table',
 'rate',
 'votes',
 'location',
 'rest_type',
 'cuisines',
 'approx_cost(for two people)',
 'listed_in(type)',
 'listed_in(city)']

In [128]:
# The lengthy approx_cost(for two people) can be shortened to 'cost2plates'
# While listed_in(type) can simply be renamed 'type'
# 'location' can also be renamed 'city'
df.rename(columns={'approx_cost(for two people)':'cost2plates', 'listed_in(type)':'serving_style', 'location':'city'},inplace=True)

### Confirm that the columns are well renamed

In [129]:
df.columns.tolist()

['name',
 'online_order',
 'book_table',
 'rate',
 'votes',
 'city',
 'rest_type',
 'cuisines',
 'cost2plates',
 'serving_style',
 'listed_in(city)']

### Checking similar columns & dropping redundant ones
- __Observing the columns above, compare 'city' and 'listed_in(city)'__
- __Load the 2 into a new dataframe for easy comparison as below__

#### Method 1: Short, using .unique() on each column individually

__Checking unique 'city' names__

In [130]:
df['city'].unique().tolist()

['Banashankari',
 'Basavanagudi',
 'Mysore Road',
 'Jayanagar',
 'Kumaraswamy Layout',
 'Rajarajeshwari Nagar',
 'Vijay Nagar',
 'Uttarahalli',
 'JP Nagar',
 'South Bangalore',
 'City Market',
 'Nagarbhavi',
 'Bannerghatta Road',
 'BTM',
 'Kanakapura Road',
 'Bommanahalli',
 nan,
 'CV Raman Nagar',
 'Electronic City',
 'HSR',
 'Marathahalli',
 'Sarjapur Road',
 'Wilson Garden',
 'Shanti Nagar',
 'Koramangala 5th Block',
 'Koramangala 8th Block',
 'Richmond Road',
 'Koramangala 7th Block',
 'Jalahalli',
 'Koramangala 4th Block',
 'Bellandur',
 'Whitefield',
 'East Bangalore',
 'Old Airport Road',
 'Indiranagar',
 'Koramangala 1st Block',
 'Frazer Town',
 'RT Nagar',
 'MG Road',
 'Brigade Road',
 'Lavelle Road',
 'Church Street',
 'Ulsoor',
 'Residency Road',
 'Shivajinagar',
 'Infantry Road',
 'St. Marks Road',
 'Cunningham Road',
 'Race Course Road',
 'Commercial Street',
 'Vasanth Nagar',
 'HBR Layout',
 'Domlur',
 'Ejipura',
 'Jeevan Bhima Nagar',
 'Old Madras Road',
 'Malleshwaram',

__Viewing unique 'listed_in(city)' names:__

In [131]:
df['listed_in(city)'].unique().tolist()

['Banashankari',
 'Bannerghatta Road',
 'Basavanagudi',
 'Bellandur',
 'Brigade Road',
 'Brookefield',
 'BTM',
 'Church Street',
 'Electronic City',
 'Frazer Town',
 'HSR',
 'Indiranagar',
 'Jayanagar',
 'JP Nagar',
 'Kalyan Nagar',
 'Kammanahalli',
 'Koramangala 4th Block',
 'Koramangala 5th Block',
 'Koramangala 6th Block',
 'Koramangala 7th Block',
 'Lavelle Road',
 'Malleshwaram',
 'Marathahalli',
 'MG Road',
 'New BEL Road',
 'Old Airport Road',
 'Rajajinagar',
 'Residency Road',
 'Sarjapur Road',
 'Whitefield']

#### Method 2: Using longer code (Checking unique vals in both columns simultaneously)

In [132]:
# Get numpy arrays with unique values from each of the 2 columns
city_columns_arrays = [df['city'].unique(), df['listed_in(city)'].unique()]
# Store column names to include in final data frame
column_names = ['city','listed_in(city)']
# Convert numpy arrays above into an iterable Pandas series using list comprehension
city_columns_arrays_series = [pd.Series(arr) for arr in city_columns_arrays]
# Concatenate the above Pandas Series into a single dataframe along col axis
city_columns_df = pd.concat(city_columns_arrays_series, axis=1, keys=column_names)
city_columns_df

city    listed_in(city)
0         Banashankari       Banashankari
1         Basavanagudi  Bannerghatta Road
2          Mysore Road       Basavanagudi
3            Jayanagar          Bellandur
4   Kumaraswamy Layout       Brigade Road
..                 ...                ...
89      West Bangalore                NaN
90         Magadi Road                NaN
91           Yelahanka                NaN
92      Sahakara Nagar                NaN
93              Peenya                NaN

[94 rows x 2 columns]

### Comparison results:
__-From the above, it is clear that both are more or less city names__
__-Hence one column shall be dropped: 'listed_in(city)'|__

In [133]:
# Check df shape first
df.shape

(51609, 11)

In [134]:
df = df.drop(['listed_in(city)'], axis=1)

In [135]:
# Check shape after
df.shape

(51609, 10)

### View DataFrame after dropping 'listed_in(city)' column

In [136]:
df.head()

name online_order book_table   rate  votes          city  \
0                  Jalsa          Yes        Yes  4.1/5    775  Banashankari   
1         Spice Elephant          Yes         No  4.1/5    787  Banashankari   
2        San Churro Cafe          Yes         No  3.8/5    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No  3.7/5     88  Banashankari   
4          Grand Village           No         No  3.8/5    166  Basavanagudi   

             rest_type                        cuisines cost2plates  \
0        Casual Dining  North Indian, Mughlai, Chinese         800   
1        Casual Dining     Chinese, North Indian, Thai         800   
2  Cafe, Casual Dining          Cafe, Mexican, Italian         800   
3          Quick Bites      South Indian, North Indian         300   
4        Casual Dining        North Indian, Rajasthani         600   

  serving_style  
0        Buffet  
1        Buffet  
2        Buffet  
3        Buffet  
4        Buffet

### Get summary info on the DataFrame: 
- After dropping columns, see remaining columns
- View non-null values in each column + data type

In [137]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51609 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   name           51609 non-null  object
 1   online_order   51609 non-null  object
 2   book_table     51609 non-null  object
 3   rate           43854 non-null  object
 4   votes          51609 non-null  int64 
 5   city           51588 non-null  object
 6   rest_type      51382 non-null  object
 7   cuisines       51564 non-null  object
 8   cost2plates    51265 non-null  object
 9   serving_style  51609 non-null  object
dtypes: int64(1), object(9)
memory usage: 4.3+ MB


## DATA CLEANING SECTION
#### NOW START CLEANING THE COLUMNS
## DATA CLEANING SECTION

## METHODS OF FINDING NULL VALUES IN COLUMNS:

### Method 1: Using inbuilt info() method and deducing:
__Running df.info() columns' non-null value counts. The column with highest null value count is one:__
 - Whose non-null value count is far off the mark from the total entries in the range index
  - Running the command below shows that 'rate' column has most null values since...
     - ...its non-null value count is 43942 while total entries is 51717  

In [138]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51609 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   name           51609 non-null  object
 1   online_order   51609 non-null  object
 2   book_table     51609 non-null  object
 3   rate           43854 non-null  object
 4   votes          51609 non-null  int64 
 5   city           51588 non-null  object
 6   rest_type      51382 non-null  object
 7   cuisines       51564 non-null  object
 8   cost2plates    51265 non-null  object
 9   serving_style  51609 non-null  object
dtypes: int64(1), object(9)
memory usage: 4.3+ MB


### Method 2: Using single line isnull().sum().idxmax() to return most nulls column:

In [139]:
most_nulls_column = df.isnull().sum().idxmax()
print(f"Column with most null values: {most_nulls_column}")

Column with most null values: rate


## Method 3: Custom function to get both column with most nulls and the count of null values:

In [140]:
max_null = 0
column_with_most_nulls = None
for col in df.columns:
    null_val_count = df[col].isnull().sum()
    if null_val_count > max_null:
        max_null = null_val_count
        column_with_most_nulls = col

print(f"Column with most null values is '{column_with_most_nulls}' and has {max_null} null values")

Column with most null values is 'rate' and has 7755 null values


### Confirming null value count in a column

In [141]:
print(f"The number of null values in rate column: {df['rate'].isnull().sum()}")

The number of null values in rate column: 7755


In [142]:
df.head()

name online_order book_table   rate  votes          city  \
0                  Jalsa          Yes        Yes  4.1/5    775  Banashankari   
1         Spice Elephant          Yes         No  4.1/5    787  Banashankari   
2        San Churro Cafe          Yes         No  3.8/5    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No  3.7/5     88  Banashankari   
4          Grand Village           No         No  3.8/5    166  Basavanagudi   

             rest_type                        cuisines cost2plates  \
0        Casual Dining  North Indian, Mughlai, Chinese         800   
1        Casual Dining     Chinese, North Indian, Thai         800   
2  Cafe, Casual Dining          Cafe, Mexican, Italian         800   
3          Quick Bites      South Indian, North Indian         300   
4        Casual Dining        North Indian, Rajasthani         600   

  serving_style  
0        Buffet  
1        Buffet  
2        Buffet  
3        Buffet  
4        Buffet

### Clean the 'rate' column

__Step 1: Check the data type__

In [143]:
type(df['rate'][0])

str

__Step 2: Check for unique values__
- To find if any outliers exist

In [144]:
df['rate'].unique().tolist()

['4.1/5',
 '3.8/5',
 '3.7/5',
 '3.6/5',
 '4.6/5',
 '4.0/5',
 '4.2/5',
 '3.9/5',
 '3.1/5',
 '3.0/5',
 '3.2/5',
 '3.3/5',
 '2.8/5',
 '4.4/5',
 '4.3/5',
 'NEW',
 '2.9/5',
 '3.5/5',
 nan,
 '2.6/5',
 '3.8 /5',
 '3.4/5',
 '4.5/5',
 '2.5/5',
 '2.7/5',
 '4.7/5',
 '2.4/5',
 '2.2/5',
 '2.3/5',
 '3.4 /5',
 '-',
 '3.6 /5',
 '4.8/5',
 '3.9 /5',
 '4.2 /5',
 '4.0 /5',
 '4.1 /5',
 '3.7 /5',
 '3.1 /5',
 '2.9 /5',
 '3.3 /5',
 '2.8 /5',
 '3.5 /5',
 '2.7 /5',
 '2.5 /5',
 '3.2 /5',
 '2.6 /5',
 '4.5 /5',
 '4.3 /5',
 '4.4 /5',
 '4.9/5',
 '2.1/5',
 '2.0/5',
 '1.8/5',
 '4.6 /5',
 '4.9 /5',
 '3.0 /5',
 '4.8 /5',
 '2.3 /5',
 '4.7 /5',
 '2.4 /5',
 '2.1 /5',
 '2.2 /5',
 '2.0 /5',
 '1.8 /5']

__Step 3: Replace "NEW" & "-" with NaN, then remove "/5" portion and return all values as floats__

In [145]:
def handle_rate(value):
    if value == '-' or value == 'NEW':
        return np.nan
    else:
        value = float(str(value).split('/')[0])
        return value

# Overwrite the existing 'rate' column
df['rate'] = df['rate'].apply(handle_rate)
df['rate'].head()

0    4.1
1    4.1
2    3.8
3    3.7
4    3.8
Name: rate, dtype: float64

__Check null values in rate column:__

In [146]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51609 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           51609 non-null  object 
 1   online_order   51609 non-null  object 
 2   book_table     51609 non-null  object 
 3   rate           41590 non-null  float64
 4   votes          51609 non-null  int64  
 5   city           51588 non-null  object 
 6   rest_type      51382 non-null  object 
 7   cuisines       51564 non-null  object 
 8   cost2plates    51265 non-null  object 
 9   serving_style  51609 non-null  object 
dtypes: float64(1), int64(1), object(8)
memory usage: 6.3+ MB


__Step 4: Fill null values, 'nan' with mean of 'rate' column__

In [147]:
# Pandas 3.0 doesn't allow this in future: df['rate'].fillna(df['rate'].mean(), inplace=True) 
# Nor this: df['rate'] = df['rate'].fillna(df['rate'].mean())
# So to opt into future behavior, I apply this method: result.infer_objects(copy=False)
# df['rate'] = df['rate'].fillna(df['rate'].mean()).infer_objects(copy=False)
# type(df['rate'])

df['rate'] = df['rate'].fillna(df['rate'].mean().round(4)) 
df['rate'].unique().tolist()

[4.1,
 3.8,
 3.7,
 3.6,
 4.6,
 4.0,
 4.2,
 3.9,
 3.1,
 3.0,
 3.2,
 3.3,
 2.8,
 4.4,
 4.3,
 3.7001,
 2.9,
 3.5,
 2.6,
 3.4,
 4.5,
 2.5,
 2.7,
 4.7,
 2.4,
 2.2,
 2.3,
 4.8,
 4.9,
 2.1,
 2.0,
 1.8]

__Step 5: Confirm all null values 'nan' have been filled with the mean__

In [148]:
# If the sum = 0, then all null values have been eliminated
df['rate'].isnull().sum()

0

### Now check DataFrame info
__- The null values have now been filled as shown in summary below__

In [149]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51609 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           51609 non-null  object 
 1   online_order   51609 non-null  object 
 2   book_table     51609 non-null  object 
 3   rate           51609 non-null  float64
 4   votes          51609 non-null  int64  
 5   city           51588 non-null  object 
 6   rest_type      51382 non-null  object 
 7   cuisines       51564 non-null  object 
 8   cost2plates    51265 non-null  object 
 9   serving_style  51609 non-null  object 
dtypes: float64(1), int64(1), object(8)
memory usage: 6.3+ MB


## 'cost2plates' column cleaning: 2nd most null values

### From df.info() results above, 'cost2plates' has most null values
__-Non-null values = 51265 while total rows = 51609__
__-Let's confirm actual null value count:__

In [159]:
df['cost2plates'].isnull().sum()

0

__Checking other unique data values in 'cost2plates' to prepare how to handle them:__

In [151]:
df['cost2plates'].unique().tolist()

['800',
 '300',
 '600',
 '700',
 '550',
 '500',
 '450',
 '650',
 '400',
 '900',
 '200',
 '750',
 '150',
 '850',
 '100',
 '1,200',
 '350',
 '250',
 '950',
 '1,000',
 '1,500',
 '1,300',
 '199',
 '80',
 '1,100',
 '160',
 '1,600',
 '230',
 '130',
 '50',
 '190',
 '1,700',
 nan,
 '1,400',
 '180',
 '1,350',
 '2,200',
 '2,000',
 '1,800',
 '1,900',
 '330',
 '2,500',
 '2,100',
 '3,000',
 '2,800',
 '3,400',
 '40',
 '1,250',
 '3,500',
 '4,000',
 '2,400',
 '2,600',
 '120',
 '1,450',
 '469',
 '70',
 '3,200',
 '60',
 '560',
 '240',
 '360',
 '6,000',
 '1,050',
 '2,300',
 '4,100',
 '5,000',
 '3,700',
 '1,650',
 '2,700',
 '4,500',
 '140']

__Removing commas from the numbers then returning floats__

In [ ]:
def handle_cost2plates(value):
    if pd.notna(value):
        return float(value.replace(',','')) if ',' in str(value) else float(value)
    else:
        return value  # Return null values as is

df['cost2plates'] = df['cost2plates'].apply(handle_cost2plates)

In [153]:
df['cost2plates'].unique().tolist()

[nan,
 1200.0,
 1000.0,
 1500.0,
 1300.0,
 1100.0,
 1600.0,
 1700.0,
 1400.0,
 1350.0,
 2200.0,
 2000.0,
 1800.0,
 1900.0,
 2500.0,
 2100.0,
 3000.0,
 2800.0,
 3400.0,
 1250.0,
 3500.0,
 4000.0,
 2400.0,
 2600.0,
 1450.0,
 3200.0,
 6000.0,
 1050.0,
 2300.0,
 4100.0,
 5000.0,
 3700.0,
 1650.0,
 2700.0,
 4500.0]

__Fill null values with column mean:__

In [154]:
df['cost2plates'] = df['cost2plates'].fillna(df['cost2plates'].mean().round(4))

__Now check null values have reduced to 0 after filling them with mean__

In [155]:
df['cost2plates'].isnull().sum()

44720

## Drop all rows with null values
#### dropna command should be run after replacing null values in numeric columns


In [156]:
df.dropna(inplace=True)

#### Check DataFrame info:
- After running the drop command above, about 600 rows are dropped comparing the info() below and before the dropna command
- Hypothesis, fill the null values in the cost for 2 plates column with average 1st then run the run dropna once more to see if dropped row count has reduced.

In [157]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6885 entries, 67 to 51716
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           6885 non-null   object 
 1   online_order   6885 non-null   object 
 2   book_table     6885 non-null   object 
 3   rate           6885 non-null   float64
 4   votes          6885 non-null   int64  
 5   city           6885 non-null   object 
 6   rest_type      6885 non-null   object 
 7   cuisines       6885 non-null   object 
 8   cost2plates    6885 non-null   float64
 9   serving_style  6885 non-null   object 
dtypes: float64(2), int64(1), object(7)
memory usage: 591.7+ KB


In [158]:
df.head(10)

name online_order book_table  rate  \
67                           Gustoes Beer House           No         No   4.1   
119                               K27 - The Pub           No         No   3.1   
166                                   Patio 805          Yes        Yes   4.1   
207                               Chutney Chang          Yes        Yes   4.1   
217                                     Toscano          Yes        Yes   4.3   
221                                       Bella          Yes        Yes   3.9   
278                 Gufha - The President Hotel          Yes        Yes   4.1   
303                         Sea Spice by 7 Star          Yes        Yes   4.0   
375                                       Salut          Yes        Yes   4.1   
397  Eurasia Pasta and Barbeque by Little Italy          Yes        Yes   4.1   

     votes          city           rest_type  \
67     868  Banashankari                 Pub   
119     30  Banashankari                 Bar   
166    364     Jayanagar       Casual Dining   
207   2339     Jayanagar       Casual Dining   
217   1129     Jayanagar       Casual Dining   
221    142     Jayanagar       Casual Dining   
278   1875     Jayanagar       Casual Dining   
303    349  Banashankari       Casual Dining   
375    442     Jayanagar  Pub, Casual Dining   
397   1063     Jayanagar  Casual Dining, Bar   

                                              cuisines  cost2plates  \
67   Continental, Italian, Mexican, North Indian, C...       1200.0   
119                              Chinese, North Indian       1000.0   
166                   Thai, Vietnamese, Asian, Chinese       1200.0   
207                         North Indian, Chinese, BBQ       1500.0   
217                                     Italian, Salad       1300.0   
221                 Mangalorean, Seafood, South Indian       1000.0   
278                     North Indian, Afghani, Mughlai       1200.0   
303  Seafood, Biryani, Beverages, South Indian, Nor...       1200.0   
375           Continental, Finger Food, Seafood, Pizza       1200.0   
397                     Italian, Fast Food, Salad, BBQ       1200.0   

    serving_style  
67       Delivery  
119      Delivery  
166      Delivery  
207      Delivery  
217      Delivery  
221      Delivery  
278      Delivery  
303      Delivery  
375      Delivery  
397      Delivery